server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java

```diff
@@ -1202,32 +1211,38 @@ public MetaData build() {
             String[] allClosedIndicesArray = allClosedIndices.toArray(new String[allClosedIndices.size()]);
 
             return new MetaData(clusterUUID, clusterUUIDCommitted, version, coordinationMetaData, transientSettings, persistentSettings,
-                    indices.build(), templates.build(), customs.build(), allIndicesArray, allOpenIndicesArray, allClosedIndicesArray,
+                    indices, templates.build(), customs.build(), allIndicesArray, allOpenIndicesArray, allClosedIndicesArray,
                     aliasAndIndexLookup);
         }
 
-        private SortedMap<String, AliasOrIndex> buildAliasAndIndexLookup() {
-            SortedMap<String, AliasOrIndex> aliasAndIndexLookup = new TreeMap<>();
-            for (ObjectCursor<IndexMetaData> cursor : indices.values()) {
-                IndexMetaData indexMetaData = cursor.value;
-                AliasOrIndex existing = aliasAndIndexLookup.put(indexMetaData.getIndex().getName(), new AliasOrIndex.Index(indexMetaData));
-                assert existing == null : "duplicate for " + indexMetaData.getIndex();
-
-                for (ObjectObjectCursor<String, AliasMetaData> aliasCursor : indexMetaData.getAliases()) {
-                    AliasMetaData aliasMetaData = aliasCursor.value;
-                    aliasAndIndexLookup.compute(aliasMetaData.getAlias(), (aliasName, alias) -> {
-                        if (alias == null) {
-                            return new AliasOrIndex.Alias(aliasMetaData, indexMetaData);
-                        } else {
-                            assert alias instanceof AliasOrIndex.Alias : alias.getClass().getName();
-                            ((AliasOrIndex.Alias) alias).addIndex(indexMetaData);
-                            return alias;
-                        }
-                    });
+        private SortedMap<String, AliasOrIndex> buildAliasAndIndexLookup(final ImmutableOpenMap<String, IndexMetaData> indices) {
+            SortedMap<String, AliasOrIndex> aliasAndIndexLookup;
+            if(!indices.equals(previousIndices)) {
```

Comment:
The house style is to prefer `== false` rather than `!`. In this case I think we can avoid this bit of controversy with this:

```
if (indices.equals(previousIndices)) {
  return previousAliasAndIndexLookup;
}
```

Then the rest of the method can be left alone.